core/src/main/java/io/questdb/cutlass/http/processors/PrometheusMetricsProcessor.java

```diff
@@ -24,42 +24,172 @@
+        final int pending = state.countPending();
+        final int wrote = r.writeBytes(state.sink.ptr() + state.written, pending);
+        state.written += wrote;
+        r.sendChunk(wrote == pending);  // Will raise `PeerIsSlowToReadException` if the tcp send buffer is full.
+    }
+
+    private void sendResponse(HttpChunkedResponseSocket r, RequestState state) throws PeerIsSlowToReadException, PeerDisconnectedException {
+        while (state.countPending() > 0) {
+            sendNextChunk(r, state);
+        }
+    }
+
+    private RequestState setupState(HttpConnectionContext context) {
+        RequestState state = LV.get(context);
+        if (state == null) {
+            state = pool.take();
+            LV.set(context, state);
+        } else {
+            state.clear();
+        }
+        return state;
+    }
+
+    /**
+     * State for processing a single request across multiple response chunks.
+     * Each object is used for the lifetime of one request, then returned the pool.
+     */
+    private static class RequestState implements QuietCloseable, Mutable {
+        /**
+         * Metrics serialization destination, sent into one or more chunks later.
+         */
+        public final DirectByteCharSink sink = new DirectByteCharSink(Files.PAGE_SIZE);
+        private final RequestStatePool pool;
+        /**
+         * Total number of bytes written to one or more chunks (`HttpChunkedResponseSocket` objects).
+         */
+        public int written = 0;
+
+        private RequestState(RequestStatePool pool) {
+            this.pool = pool;
+        }
+
+        @Override
+        public void clear() {
+            sink.clear();
+            written = 0;
+        }
+
+        /**
+         * Return to pool at the end of a request.
+         */
+        @Override
+        public void close() {
+            clear();
+            pool.give(this);
+        }
+
+        /**
+         * Release off-heap buffer.
+         */
+        public void free() {
+            sink.close();
+        }
+
+        /**
+         * Calculate the number of bytes that still need to be written to chunks.
+         */
+        public int countPending() {
+            return sink.size() - written;
+        }
+    }
+
+    public static class RequestStatePool implements QuietCloseable {
+        private final ObjList<RequestState> objects;
+
+        public RequestStatePool() {
+            objects = new ObjList<>();
+        }
+
+        @TestOnly
+        public synchronized int size() {
+            return objects.size();
+        }
+
+        @Override
+        public void close() {
+            for (int i = 0, n = objects.size(); i < n; i++) {
+                objects.getQuick(i).free();
+            }
+            objects.clear();
+        }
+
+        public synchronized void give(RequestState requestState) {
+            objects.add(requestState);
```

Comment:
There should be a max size ideally after which we close the incoming `requestState` instead of pooling all instances forever.

Reply:
I've implemented this, capped to the worker count.